File: ios/LibreStage/Core/SongDuplicateCheck.swift
// libre-stage iOS App
// Copyright (C) 2026 libre-stage contributors
// SPDX-License-Identifier: GPL-3.0-or-later

import Foundation

struct SongDuplicateMatch {
    let score: Double
    let song: SongOut
}

enum SongDuplicateCheck {
    static func findBestSongDuplicate(candidate: SongDetailsDraft, songs: [SongOut]) -> SongDuplicateMatch? {
        let candidateTitle = normalizeSongText(candidate.title)
        let candidateInterpret = normalizeSongText(candidate.interpret)

        guard !candidateTitle.isEmpty, !candidateInterpret.isEmpty else {
            return nil
        }

        var bestMatch: SongDuplicateMatch?

        for song in songs {
            let songTitle = normalizeSongText(song.title)
            let songInterpret = normalizeSongText(song.interpret)
            guard !songTitle.isEmpty, !songInterpret.isEmpty else {
                continue
            }

            let exact = candidateTitle == songTitle && candidateInterpret == songInterpret
            let titleScore = bestFieldScore(candidateTitle, songTitle)
            let interpretScore = bestFieldScore(candidateInterpret, songInterpret)
            let combinedScore = exact ? 1 : titleScore * 0.6 + interpretScore * 0.4

            let passesBalancedThreshold =
                exact ||
                (titleScore >= 0.84 && interpretScore >= 0.84) ||
                combinedScore >= 0.88

            guard passesBalancedThreshold else {
                continue
            }

            if bestMatch == nil || combinedScore > (bestMatch?.score ?? 0) {
                bestMatch = SongDuplicateMatch(score: combinedScore, song: song)
            }
        }

        return bestMatch
    }

    static func normalizeSongText(_ value: String?) -> String {
        let base = (value ?? "")
            .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "en_US_POSIX"))

        let alnumSeparated = base.replacingOccurrences(
            of: "[^a-z0-9]+",
            with: " ",
            options: .regularExpression
        )

        return alnumSeparated
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
    }

    private static func sortTokens(_ value: String) -> String {
        normalizeSongText(value)
            .split(separator: " ")
            .map(String.init)
            .sorted()
            .joined(separator: " ")
    }

    private static func similarity(_ left: String, _ right: String) -> Double {
        if left.isEmpty || right.isEmpty {
            return 0
        }
        if left == right {
            return 1
        }

        let maxLen = max(left.count, right.count)
        if maxLen == 0 {
            return 1
        }

        let distance = levenshteinDistance(left, right)
        return 1 - (Double(distance) / Double(maxLen))
    }

    private static func bestFieldScore(_ left: String, _ right: String) -> Double {
        let direct = similarity(normalizeSongText(left), normalizeSongText(right))
        let tokenSorted = similarity(sortTokens(left), sortTokens(right))
        return max(direct, tokenSorted)
    }

    private static func levenshteinDistance(_ left: String, _ right: String) -> Int {
        if left == right { return 0 }
        if left.isEmpty { return right.count }
        if right.isEmpty { return left.count }

        let leftChars = Array(left)
        let rightChars = Array(right)

        var previous = Array(0...rightChars.count)
        var current = Array(repeating: 0, count: rightChars.count + 1)

        for (i, leftChar) in leftChars.enumerated() {
            current[0] = i + 1
            for (j, rightChar) in rightChars.enumerated() {
                let cost = leftChar == rightChar ? 0 : 1
                current[j + 1] = min(
                    previous[j + 1] + 1,
                    current[j] + 1,
                    previous[j] + cost
                )
            }
            swap(&previous, &current)
        }

        return previous[rightChars.count]
    }
}

File: frontend/src/lib/songDuplicateCheck.js
/**
 * Normalisiert Strings fuer robusten Textvergleich.
 * - Lowercase
 * - Akzente entfernen
 * - Sonderzeichen zu Leerzeichen
 * - Mehrfach-Leerzeichen reduzieren
 */
export function normalizeSongText(value) {
  return String(value ?? '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, ' ')
    .trim()
    .replace(/\s+/g, ' ');
}

function sortTokens(value) {
  return normalizeSongText(value)
    .split(' ')
    .filter(Boolean)
    .sort()
    .join(' ');
}

function levenshteinDistance(a, b) {
  if (a === b) return 0;
  if (!a.length) return b.length;
  if (!b.length) return a.length;

  const prev = new Array(b.length + 1);
  const curr = new Array(b.length + 1);

  for (let j = 0; j <= b.length; j += 1) prev[j] = j;

  for (let i = 1; i <= a.length; i += 1) {
    curr[0] = i;
    for (let j = 1; j <= b.length; j += 1) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(
        prev[j] + 1,
        curr[j - 1] + 1,
        prev[j - 1] + cost
      );
    }
    for (let j = 0; j <= b.length; j += 1) prev[j] = curr[j];
  }

  return prev[b.length];
}

function similarity(a, b) {
  if (!a || !b) return 0;
  if (a === b) return 1;
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1;
  return 1 - levenshteinDistance(a, b) / maxLen;
}

function bestFieldScore(left, right) {
  const direct = similarity(normalizeSongText(left), normalizeSongText(right));
  const tokenSorted = similarity(sortTokens(left), sortTokens(right));
  return Math.max(direct, tokenSorted);
}

/**
 * Findet den besten Dubletten-Treffer fuer die Kombination aus Titel+Interpret.
 * Liefert null, wenn keine ausreichend gute Uebereinstimmung gefunden wurde.
 */
export function findBestSongDuplicate(candidate, songs = []) {
  const candidateTitle = normalizeSongText(candidate?.title);
  const candidateInterpret = normalizeSongText(candidate?.interpret);

  if (!candidateTitle || !candidateInterpret) return null;

  let bestMatch = null;

  for (const song of songs) {
    const songTitle = normalizeSongText(song?.title);
    const songInterpret = normalizeSongText(song?.interpret);
    if (!songTitle || !songInterpret) continue;

    const exact = candidateTitle === songTitle && candidateInterpret === songInterpret;
    const titleScore = bestFieldScore(candidateTitle, songTitle);
    const interpretScore = bestFieldScore(candidateInterpret, songInterpret);
    const combinedScore = exact ? 1 : titleScore * 0.6 + interpretScore * 0.4;

    const passesBalancedThreshold =
      exact ||
      (titleScore >= 0.84 && interpretScore >= 0.84) ||
      combinedScore >= 0.88;

    if (!passesBalancedThreshold) continue;

    if (!bestMatch || combinedScore > bestMatch.score) {
      bestMatch = {
        score: combinedScore,
        song
      };
    }
  }

  return bestMatch;
}
